nox/sessions.py

```diff
@@ -368,6 +374,15 @@ def conda_install(
         if not args:
             raise ValueError("At least one argument required to install().")
 
+        if self._runner.global_config.no_install:
+            if venv._venv_created:
+                logger.info("Skipping conda installation, as --no-install is set.")
+                return None
+            else:
+                logger.info(
+                    "Venv not created yet. Ignoring --no-install and installing from conda."
+                )
```

Comment:
I think the condition is inverted here. If Nox created the virtualenv afresh, then we should *not* skip installation.

```suggestion
            if venv._venv_created:
                logger.info(
                    "Venv not created yet. Ignoring --no-install and installing from conda."
                )
            else:
                logger.info("Skipping conda installation, as --no-install is set.")
                return None
```

Reply:
Yes, I see what you're saying. 

I have been conceptualising the goal as using `ProcessEnv._venv_created` to determine whether to skip `session.install`. However, I had been misthinking of nox as a stateful app, and I had only been considering the _first_ run of a set of nox runs. Reading your suggestion, I think that it's runs 2...N that we're actually trying to optimise.

That's to say, what we're trying to do is figure out if the venvs already exist or not.  Now that I re-read section 2 of the comment at https://github.com/theacodes/nox/issues/394#issuecomment-799776880, you had said the problem this variable is addressing is "what we should do if an environment does not exist yet. ", so I think you had understood this earlier than I had.

What you're suggesting here makes sense, but I think it's showing that _both the variable name I had picked is a bad one, _and_ that the variable is tracking the wrong concept_ 🙂. I think we're not interested in whether a venv was created in this session, we're interested in whether a venv was created in _any_ run of nox.

The `--reuse-existing-virtualenvs` flag solves this by checking the file system to see whether the venv already exists before it attempts a delete. This is from virtualenv.py line 198 in my branch:

```
    def _clean_location(self) -> bool:
        """Deletes existing conda environment"""
        if os.path.exists(self.location): #<--- checking for state!
            if self.reuse_existing:
                return False
            else:
                cmd = ["conda", "remove", "--yes", "--prefix", self.location, "--all"]
```

Should `_venv_created` maybe have its implementation changed to something like:

```
self._venv_created = os.path.exists(self.location)
```

This way, it's checking the state generated by previous runs of nox, instead of just assuming that a venv was not created when it's initialised. With that change, I think my current implementation would make sense.

What do you think?

Reply:
Just in case it got lost in the to and fro of this review, I'll repeat my earlier suggestion from https://github.com/theacodes/nox/issues/394#issuecomment-801789356:

> The virtualenv needs to remember if it was reused. This could be a bool attribute `ProcessEnv._reused`, initially False. The VirtualEnv and CondaEnv implementations can set this to True in their `create` methods when the virtualenv gets reused. The logic is already there: These methods call `self._clean_location` which returns False if the virtualenv was reused.

(Edited to rename `reused` to `_reused` to avoid adding to the public API.)

Personally, I think the name `reused` avoids precisely the misunderstanding you described in your post above.

> Should `_venv_created` maybe have its implementation changed to something like:
>
> ```
> self._venv_created = os.path.exists(self.location)
> ```

Maybe I'm missing something here, but I don't see how this would work. I presume you want this code before we create the virtualenv, i.e. at the start of `VirtualEnv.create` and `CondaEnv.create`. But even if the virtualenv exists then, it may get removed and created from scratch. Instead you need the check that is already done inside `_clean_location`: `os.path.exists(self.location) and self.reuse_existing`. Instead of copying this check, I would recommend to use the return value of `_clean_location`, as described above.

Reply:
I'm went ahead with the renaming from _venv_created to _reused because it was simple, but I'll come back to this later to finish up your other suggestions.

You were right, I had missed the advice around `self._clean_location`. Thanks for bringing it back up!

Reply:
@cjolowicz Okay! I renamed `_venv_created` to `_reused` and then moved `_reused` to be behind the check of `_clean_location`. 

I simplified the tests a little as well, because I realised some of them were redundant.

When I read the code, it reads to me like the check for `--no-install` and `_reused` are doing the "right" things, but I made that mistake last time, so I may be mistaken 😞

I can see the comments in https://github.com/theacodes/nox/issues/394 about maybe renaming this from `--no-install` to `--reuse-install`. Should I close this PR while that conversation gets resolved? Or leave this open?